src/network/network_gui.cpp

```diff
@@ -1806,166 +1793,543 @@ static void PopupClientList(ClientID client_id, int x, int y)
+					NWidget(NWID_SPACER), SetMinimalSize(20, 0), SetFill(1, 0), SetResize(1, 0),
+					NWidget(WWT_DROPDOWN, COLOUR_GREY, WID_CL_SERVER_VISIBILITY), SetDataTip(STR_BLACK_STRING, STR_NETWORK_CLIENT_LIST_SERVER_VISIBILITY_TOOLTIP),
+				EndContainer(),
+			EndContainer(),
+		EndContainer(),
+		NWidget(WWT_FRAME, COLOUR_GREY), SetDataTip(STR_NETWORK_CLIENT_LIST_PLAYER, STR_NULL), SetPadding(0, 4, 4, 4), SetPIP(0, 2, 0),
+			NWidget(NWID_HORIZONTAL), SetPIP(0, 3, 0),
+				NWidget(WWT_TEXT, COLOUR_GREY), SetMinimalSize(0, 12), SetDataTip(STR_NETWORK_CLIENT_LIST_PLAYER_NAME, STR_NULL),
+				NWidget(NWID_SPACER), SetMinimalSize(20, 0),
+				NWidget(WWT_EMPTY, COLOUR_GREY, WID_CL_CLIENT_NAME), SetFill(1, 0), SetResize(1, 0),
+				NWidget(WWT_PUSHIMGBTN, COLOUR_GREY, WID_CL_CLIENT_NAME_EDIT), SetMinimalSize(12, 14), SetDataTip(SPR_RENAME, STR_NETWORK_CLIENT_LIST_PLAYER_NAME_EDIT_TOOLTIP),
+			EndContainer(),
+		EndContainer(),
+		NWidget(NWID_HORIZONTAL),
+			NWidget(WWT_MATRIX, COLOUR_GREY, WID_CL_MATRIX), SetMinimalSize(180, 0), SetResize(1, 1), SetFill(1, 1), SetMatrixDataTip(1, 0, STR_NULL), SetScrollbar(WID_CL_SCROLLBAR),
+			NWidget(NWID_VERTICAL),
+				NWidget(NWID_VSCROLLBAR, COLOUR_GREY, WID_CL_SCROLLBAR),
+				NWidget(WWT_RESIZEBOX, COLOUR_GREY),
+			EndContainer(),
+		EndContainer(),
+	EndContainer(),
 };
 
 static WindowDesc _client_list_desc(
-	WDP_AUTO, "list_clients", 0, 0,
+	WDP_AUTO, "list_clients", 220, 300,
 	WC_CLIENT_LIST, WC_NONE,
 	0,
 	_nested_client_list_widgets, lengthof(_nested_client_list_widgets)
 );
 
+typedef void (*ButtonCompanyCallback)(class NetworkClientListWindow *w, Point pt, CompanyID company_id);
+typedef void (*ButtonClientCallback)(class NetworkClientListWindow *w, Point pt, ClientID client_id);
+
+/**
+ * Button shown for either a company or client in the client-list.
+ *
+ * These buttons are dynamic and strongly depends on which company/client
+ * what buttons are available. This class allows dynamically creating them
+ * as the current Widget system does not.
+ */
+struct Button {
+	StringID label;                      ///< The label of the button.
+	StringID tooltip;                    ///< The tooltip of the button.
+	Colours colour;                      ///< The colour of the button.
+	uint height;                         ///< Calculated height of the button.
+	uint width;                          ///< Calculated width of the button.
+	CompanyID company_id;                ///< Company ID this button belongs to.
+	ButtonCompanyCallback company_proc;  ///< Callback proc to call when button is pressed, or nullptr if not a company button.
+	ClientID client_id;                  ///< Client ID this button belongs to.
+	ButtonClientCallback client_proc;    ///< Callback proc to call when button is pressed, or nullptr if not a client button.
+
+	Button(StringID label, StringID tooltip, Colours colour, CompanyID company_id, ButtonCompanyCallback proc) :
```

Comment:
I'm wondering whether this can be less duplicated, as now the button caller needs to figure out whether to call the company or client callback, or do you later need a button to perform the call back om both a company and client at the same button click?

Reply:
No, but I wanted to keep the callbacks pure, that it is clear if they are either a Company or a Client callback. A button will never be used for both (that would also make little sense). So I opted for 2 ctors and 2 function prototypes to ensure the callbacks need to be either CompanyID or ClientID based.

There might be other C++ solutions here, but .. I am not aware of those (which says little, my C++ knowledge is not that extensive :D)

Reply:
Maybe with a template.

Reply:
Yeah, such comments are not helping me at all :) If that would be possible / useful, I would need a bit more guidance than just mentioning C++ has templates :D (sorry :P)

Reply:
@glx22: tnx for the example :D (handled via IRC). That really helped :)